Add loading indicator to "Schedule" button.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
- First commit is for refactoring the existing code.
- Second commit adds loading indicator to "Schedule" button.
Fixes: <!-- Issue link, or clear description.--> #25182

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
![schedule-msg-loader](https://user-images.githubusercontent.com/35494118/234242473-9d9a9878-fe4f-4575-85d8-79a9be2af222.gif)


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented on web/src/compose_banner.ts:
```diff
@@ -52,9 +52,9 @@ export function clear_message_sent_banners(): void {
-    $("#compose-send-button").removeClass("disable-btn");
+    $(".compose_message_button .loader").hide();
+    $(".compose_message_button span").show();
+    $(".compose_message_button").removeClass("disable-btn");
```
Maybe `compose-submit-button` would be a better name? I think that would make the role clear. (I think we're also preferring `-`s over `_`s for separators in CSS).

sahil839 replied:
Updated it.

sahil839 commented on web/src/reminder.js:
```diff
@@ -127,6 +127,7 @@ export function schedule_message(
     if (scheduled_message_id) {
         request.scheduled_message_id = scheduled_message_id;
     }
+    compose_ui.show_compose_spinner();
```
There were some bugs for the spinner not hiding correctly, when using slash commands. Added a commit to fix it.

sahil839 commented:
Updated the PR -
- Fixed the new class name as suggested above.
- Added commit to hide the spinner in case of success and error when using slash command to schedule messages. Also, added code to reset the compose box. i.e. enable the textarea element and clear it, after successfully scheduling the message using slash command.

There is still a minor thing to fix where the topic input is focussed after scheduling message using keyboard and not by clicking on button using mouse. This is probably due to the textarea element being disabled when focus is triggered on it. Will try to fix the focus thing by tommorrow.

Related to this, the comments in code mention that we disable the textarea element when using slash command, but we also do it when scheduling message by clicking on the button normally since both use same functions. Do we want to disable the button in both cases?

alya commented:
I think disabling the textarea while the message is in the process of being scheduled makes sense.

Note that #25240 will remove the slash command.

timabbott commented:
> I think disabling the textarea while the message is in the process of being scheduled makes sense.

Yeah I think in general we should just copy the UI patterns we have for sending a message that cannot be locally echoed.

timabbott commented:
This is a great checkpoint, merged, thanks @sahil839! We can fix the focus issue you noted in a next PR; I wanted to get this in just so future work can have this piece of the implementation already there.
